### PR TITLE
Don't pass the `Type` but use given `Type[T]` & `Quotes`

### PR DESCRIPTION
**Pull request by saeltz**
This fixes some compilation errors in the Scala 3 branch. Of course not all of them. But more things are compiling now.

#### Has the version number been increased?
 - [ ] Yes! (or not but it was intended to not increase it)

### CODE REVIEW

File: src/main/scala/io/moia/protos/teleproto/FormatImpl.scala [hunk 1]
```diff
@@ -71,42 +72,42 @@ trait FormatImpl {
...
-        cs.isCaseClass && {
-          val cons = tpe.member(termNames.CONSTRUCTOR).asMethod
-          cons.paramLists.lengthCompare(1) == 0 // has a single parameter list
-        }
-      case _ => false
+  private[teleproto] def isSimpleCaseClass[T](tpe: Type[T])(using Quotes): Boolean = {
+    import quotes.reflect.*
+
+    val tpeSym = TypeTree.of[T].symbol
+    tpeSym.flags.is(Flags.Case) && {
```
Comment (saeltz) [inline src/main/scala/io/moia/protos/teleproto/FormatImpl.scala:106]: This should be it already. I don't know what the second part is doing and left it untouched.

File: src/main/scala/io/moia/protos/teleproto/FormatImpl.scala [hunk 2]
```diff
@@ -71,42 +72,42 @@ trait FormatImpl {
...
-      case cs: ClassSymbol =>
-        cs.isCaseClass && {
-          val cons = tpe.member(termNames.CONSTRUCTOR).asMethod
-          cons.paramLists.lengthCompare(1) == 0 // has a single parameter list
-        }
-      case _ => false
+  private[teleproto] def isSimpleCaseClass[T](tpe: Type[T])(using Quotes): Boolean = {
+    import quotes.reflect.*
+
+    val tpeSym = TypeTree.of[T].symbol
```
Comment (saeltz) [inline src/main/scala/io/moia/protos/teleproto/FormatImpl.scala:105]: This is how to get the `symbol` in Scala 3. The `tpe` could also be `given` I think, as we don't access the parameter directly.

File: src/main/scala/io/moia/protos/teleproto/FormatImpl.scala [hunk 3]
```diff
@@ -137,22 +138,22 @@ trait FormatImpl {
...
 
-  private[teleproto] def isSealedTrait(tpe: Type): Boolean =
-    tpe.typeSymbol match {
-      case cs: ClassSymbol => cs.isSealed && cs.isAbstract
-      case _               => false
-    }
+  private[teleproto] def isSealedTrait[T](tpe: Type[T])(using Quotes): Boolean = {
+    import quotes.reflect.*
+
+    TypeTree.of[T].symbol.flags.is(Flags.Trait & Flags.Sealed)
```
Comment (saeltz) [inline src/main/scala/io/moia/protos/teleproto/FormatImpl.scala:144]: This works.

File: src/main/scala/io/moia/protos/teleproto/ReaderImpl.scala [hunk 4]
```diff
@@ -16,23 +16,15 @@
 
 package io.moia.protos.teleproto
 
-import scala.reflect.macros.blackbox
```
Comment (saeltz) [inline src/main/scala/io/moia/protos/teleproto/ReaderImpl.scala:19]: That whole blackbox macro doesn't exist anymore in Scala 3.

File: src/main/scala/io/moia/protos/teleproto/ReaderImpl.scala [hunk 5]
```diff
@@ -16,23 +16,15 @@
...
 import scala.compiletime.error
+import scala.quoted.*
 
-class ReaderImpl(val c: blackbox.Context) extends FormatImpl {
-  import c.universe._
```
Comment (saeltz) [inline src/main/scala/io/moia/protos/teleproto/ReaderImpl.scala:23]: Thus we cannot import the context.

File: src/main/scala/io/moia/protos/teleproto/FormatImpl.scala (hunk 2 above)
Comment (willbinzi) [reply src/main/scala/io/moia/protos/teleproto/FormatImpl.scala:105]: I think you're right. I think we can just specify `[T: Type]` in the type parameterisation? I think we're acutally going to be using this `tpeSim` most of the time instead, right?

File: src/main/scala/io/moia/protos/teleproto/FormatImpl.scala (hunk 1 above)
Comment (willbinzi) [reply src/main/scala/io/moia/protos/teleproto/FormatImpl.scala:106]: I think it's checking that the constructor takes only 1 argument? i.e. the case class is just a wrapper around another value (presumably what they mean by `simple`?)

File: src/main/scala/io/moia/protos/teleproto/FormatImpl.scala (hunk 3 above)
Comment (willbinzi) [reply src/main/scala/io/moia/protos/teleproto/FormatImpl.scala:144]: Nice!

File: src/main/scala/io/moia/protos/teleproto/ReaderImpl.scala [hunk 6]
```diff
@@ -61,10 +53,10 @@ class ReaderImpl(val c: blackbox.Context) extends FormatImpl {
     *
     * The result is `f` applied to the reader expression with the (possible) compatibility issues of reader generation (if happened).
     */
-  private def withImplicitReader(protobufType: Type, modelType: Type)(compileInner: Tree => Tree): Compiled = {
+  private def withImplicitReader[P, M](protobufType: Type[P], modelType: Type[M])(compileInner: Tree => Tree)(using Quotes): Compiled = {
```
Comment (willbinzi) [inline src/main/scala/io/moia/protos/teleproto/ReaderImpl.scala:56]: I think at this point we should be working with `TypeRepr` instead of `Type`, right?

**Pull request merged**